File: modules/rust_workloads/benches/bench_matmul.rs
//! Matrix-multiply bench: the `ijk` vs `ikj` loop-order gap is cache
//! behaviour — in-process profiles show a hot loop, an external sampler's
//! cache counters explain the difference.

use std::hint::black_box;

use criterion::{Criterion, criterion_group, criterion_main};
use pprof::criterion::{Output, PProfProfiler};
use rust_workloads::{matmul, workload_n};

const DEFAULT_N: usize = 256;

fn bench_matmul(c: &mut Criterion) {
    let n = workload_n(DEFAULT_N);
    let a = matmul::random_matrix(n, 42);
    let b = matmul::random_matrix(n, 43);
    let mut group = c.benchmark_group("matmul");
    group.bench_function("ijk", |bench| {
        bench.iter(|| matmul::multiply_ijk(black_box(&a), black_box(&b), n));
    });
    group.bench_function("ikj", |bench| {
        bench.iter(|| matmul::multiply_ikj(black_box(&a), black_box(&b), n));
    });
    group.finish();
}

criterion_group! {
    name = benches;
    config = Criterion::default().with_profiler(PProfProfiler::new(100, Output::Protobuf));
    targets = bench_matmul
}
criterion_main!(benches);

File: modules/java_workloads/benches/BenchMatmul.java
// Matrix-multiply bench: the ijk vs ikj loop-order gap is cache
// behaviour — in-process profiles show a hot loop, an external
// sampler's cache counters explain the difference.

package monorepo.workloads;

import java.util.concurrent.TimeUnit;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;

@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Benchmark)
@Fork(1)
public class BenchMatmul {
    private static final int DEFAULT_N = 256;

    private int n;
    private double[] a;
    private double[] b;

    @Setup
    public void setup() {
        n = WorkloadN.workloadN(DEFAULT_N);
        a = Matmul.randomMatrix(n, 42);
        b = Matmul.randomMatrix(n, 43);
    }

    @Benchmark
    public double[] ijk() {
        return Matmul.multiplyIjk(a, b, n);
    }

    @Benchmark
    public double[] ikj() {
        return Matmul.multiplyIkj(a, b, n);
    }
}
